README.md
stm32l0xx-hal
=============

[![Build Status](https://travis-ci.com/stm32-rs/stm32l0xx-hal.svg?branch=master)](https://travis-ci.com/stm32-rs/stm32l0xx-hal)

> :warning: **Note:** The stm32l0xx HAL team is looking for more maintainers!
> Do you think you could help out with reviewing contributions and improving
> the codebase? If yes, please let us know in issue
> [#211](https://github.com/stm32-rs/stm32l0xx-hal/issues/211)!

[_stm32l0xx-hal_](https://github.com/stm32-rs/stm32l0xx-hal) is a Hardware
Abstraction Layer (HAL) for the STMicro STM32L0xx family of microcontrollers.

This crate relies on Adam Greig's [stm32l0](https://crates.io/crates/stm32l0)
crate to provide appropriate register definitions and implements a partial set
of the [embedded-hal](https://github.com/rust-embedded/embedded-hal) traits.

Based on the [stm32l1xx-hal](https://github.com/stm32-rs/stm32l1xx-hal) crate
by Vitaly Domnikov and the [stm32f4xx-hal](https://github.com/stm32-rs/stm32f4xx-hal)
crate by Daniel Egger.


# Usage

Add the [`stm32l0xx-hal` crate](https://crates.io/crates/stm32l0xx-hal) to your
dependencies in `Cargo.toml` and make sure to pick the appropriate `mcu-*`
Cargo feature to enjoy the full feature set for your MCU (see next section
"Supported Configurations" for more details).

For example, when using the STM32L071KBTx MCU:

```toml
[dependencies]
stm32l0xx-hal = { version = "0.10.0", features = ["mcu-STM32L071KBTx", "rt"] }
```


# Supported Configurations

The STM32L0 family consists of different subfamilies with different peripherals
and I/O configurations. Superficially, the family can be grouped into the
groups `stm32l0x1`, `stm32l0x2` and `stm32l0x3`. However, some aspects like
alternate function mappings for I/O pins do not follow these groups.

In order for the HAL to properly support all those MCUs, we generate some
peripheral mappings and corresponding Cargo features using
[cube-parse](https://github.com/stm32-rs/cube-parse/).

## MCU Features (`mcu-*`)

The easiest way for you to get started, is to use your appropriate `mcu-*`
feature. For example, when using the STM32L071KBTx MCU, you just set the
`mcu-STM32L071KBTx` feature in `Cargo.toml`:

```toml
# Cargo.toml
[dependencies]
stm32l0xx-hal = { version = "0.10.0", features = ["mcu-STM32L071KBTx", "rt"] }
```

If you take a look at the [`Cargo.toml`
file](https://github.com/stm32-rs/stm32l0xx-hal/blob/master/Cargo.toml), you
can see that `mcu-STM32L071KBTx` is just an alias for `["io-STM32L071",
"eeprom-6144", "flash-128", "ram-20"]`.

## I/O Features (`io-*`)

The `io-*` features are based on the GPIO peripheral version. This determines
the pin function mapping of the MCU. The features seem to correspond to the
product categories.

Right now, the following features are supported:

- `io-STM32L021` (Product category 1)
- `io-STM32L031` (Product category 2)
- `io-STM32L051` (Product category 3)
- `io-STM32L071` (Product category 5)

The product categories should be listed in your MCU family datasheet. The name
of the `io-*` feature itself is derived from the internal name used in the
STM32CubeMX database. It does not necessarily match the name of the MCU,
for example the `STM32L062K8Tx` uses the GPIO peripheral version named
`io-STM32L051`.

## Memory Layout: Flash (`flash-*`) and RAM (`ram-*`)

Flash and RAM sizes are automatically configured by using the appropriate
`mcu-*` Cargo feature, as described further up.

If you have an MCU that does not yet have a corresponding `mcu-*` feature, or
if the default configuration is incorrect, you can override the `memory.x` of
`stm32l0xx-hal` by providing your own:

In your crate root, add a file called `memory.x` with the correct
configuration. For example:

```
MEMORY
{
  FLASH : ORIGIN = 0x08000000, LENGTH = 64K
  RAM : ORIGIN = 0x20000000, LENGTH = 8K
}
```

Add a `build.rs` file with the following content:

```rust
use std::env;
use std::fs::File;
use std::io::Write;
use std::path::PathBuf;

fn main() {
    // Put the linker script somewhere the linker can find it
    let out = &PathBuf::from(env::var_os("OUT_DIR").unwrap());
    File::create(out.join("memory.x"))
        .unwrap()
        .write_all(include_bytes!("memory.x"))
        .unwrap();
    println!("cargo:rustc-link-search={}", out.display());

    println!("cargo:rerun-if-changed=build.rs");
    println!("cargo:rerun-if-changed=memory.x");
}
```

And finally add the `disable-linker-script` feature to your `stm32l0xx-hal` dependency:

```toml
# Cargo.toml
[dependencies]
stm32l0xx-hal = { version = "0.10.0", features = ["mcu-STM32L071K8Ux", "disable-linker-script"] }
```

# Other Cargo Features

- `rtc`: Enable the RTC API based on [rtcc](https://docs.rs/rtcc/). Enabled by default.

# Toolchain Setup

In order to use this HAL, you need the following Setup:

1. Install Rustup

    See [rustup.rs](https://rustup.rs/) for details. You may als be able to
    install Rustup directly through your distro.

2. Install the `arm-none-eabi` compiler toolchain

	https://developer.arm.com/open-source/gnu-toolchain/gnu-rm/downloads

    If you cannot install the toolchain directly through your OS / distro, we
    recommend installing the precompiled binaries to '/usr/local/opt'.  Add the
    bin folders (/bin & /arm-none-eabi/bin) to your environments variable 'PATH'.

3. Install the `thumbv6m-none-eabi` target for Rust

    Simply run `rustup target add thumbv6m-none-eabi`

4. Install [probe-run](https://crates.io/crates/probe-run) to run examples.

    ```bash
    cargo install probe-run
    ```

For more instructions on how to get started with ARM / Cortex-M programming
using Rust, check out the [Embedded Rust
Book](https://rust-embedded.github.io/book/).


# Build Examples

You can build examples through Cargo:

    $ cargo build --release --examples --features stm32l0x1,rt

Note that not all examples are compatible with all MCUs. You might need to peek
into the example source code.

# Run Examples

This crate uses [probe-run](https://crates.io/crates/probe-run) to run examples on target hardware. 

To run the `blinky` example on an STM32L053 Nucleo:

```bash
cargo run --example blinky --features mcu-STM32L053R8Tx -- --chip STM32L053R8Tx
```

# Flashing Using Helper Scripts

The following instructions outline how-to on flashing the 'serial' example
code. This can be extended to any other example code.

## Flashing with ST-Flash:

1. Flash the microcontroller using the flash script
    ``` 
    $ ./flash.sh target/thumbv6m-none-eabi/release/examples/serial
    ```

## Flashing with OpenOCD

1. Flash the microcontroller using the openocd flash script
    ``` 
    $ ./openocd_flash.sh target/thumbv6m-none-eabi/release/examples/serial
    ```

# Contributor Notes

- Revert local dependencies to external cargo and uncomment configurations
  before committing


# License

0-Clause BSD License, see [LICENSE-0BSD.txt](LICENSE-0BSD.txt) for more details.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/).

## [Unreleased]

<!-- When making a PR, please update this section. Note: This document should
make upgrading the HAL as painless as possible! If it makes sense, feel free
to add upgrade notes and examples. When adding an issue or PR reference, don't
forget to update the links at the bottom of the changelog as well.-->

### Additions

- SPI: Allow output pins of mode `PushPull` and `OpenDrain` ([#226])
- Enable TIM2 outputs on `PA5`, `PA15`, `PB3` for all devices in the L0 family (previously only 0x2
  and 0x3) ([#224])
- Enable SPI2 on subset of stm32l0x1 devices ([#221])
- Add `pause` and `resume` methods to timers ([#220])

### Breaking Changes

- Allow selection of RTC clock source to be LSE or LSI ([#218])

### Non-Breaking Changes

### Fixes

- Correct calculation of AHB prescaler for factors > 16 ([#228])

### Documentation

## [v0.10.0] - 2022-08-15

### Additions

### Breaking Changes

- Upgrade `stm32l0` to latest version, and other dependency updates ([#215]):

  `dependencies`

  | crate       | from   | to     |
  | ----------- | ------ | ------ |
  | cast        | 0.2.2  | 0.3.0  |
  | cortex-m-rt | 0.6.8  | 0.7.0  |
  | rtcc        | 0.2    | 0.3.0  |
  | stm32l0     | 0.13.0 | 0.15.1 |

  `dev-dependencies`

  | crate                | from  | to    |
  | -------------------- | ----- | ----- |
  | aligned              | 0.3.1 | 0.4.1 |
  | cortex-m-rtic        | 0.5.6 | 1.1.3 |
  | cortex-m-semihosting | 0.3.2 | 0.5.0 |
  | panic-semihosting    | 0.5.1 | 0.6.0 |

### Non-Breaking Changes

### Fixes

- Add other possible pins for SPI1 for L0x1 family

### Documentation

## [v0.9.0] - 2021-12-11

### Additions

- Add `is_idle`, `is_rx_not_empty` and `clear_idle_interrupt` methods to serial ([#206])

### Breaking Changes

- Unify RCC enable / reset ([#196])
- Setup `hsi16diven` when running `Rcc::freeze` ([#197])

### Non-Breaking Changes

- Fix clippy lints ([#200])
- Make cortex-m-semihosting a dev dependency ([#208])

### Documentation

- Fix Cargo features used by docs.rs ([#203])

## [v0.8.0] - 2021-11-04

### Additions

- Add `Timer::new` to create timer without starting. ([#152])
- Add (untested) example `temperature` ([#161]) which uses `adc` to read the internal temperature
  and also an externally connected TMP36 analog sensor.
- Add `Pwm::set_frequency` to allow dynamically changing the underlying PWM timer's frequency. This
  was previously impossible without resorting to `unsafe` because the channels where moved out of
  the timer struct and Rust's ownership rules forbade us to borrow the timer to call
  `set_frequency`. ([#174])
- Add Cargo features for flash and RAM sizes allowing to pass the correct sizes to the linker
  instead of having a default for the whole sub-family. ([#173])

### Breaking Changes

- Migrate from custom `Hertz` implementation to
  [`embedded-time`](https://crates.io/crates/embedded-time) ([#183])
- Add `enable` to `GeneralPurposeTimer`
- `Instance::clock_frequency` is now an associated function and doesn't take `&self` anymore.
- Build script requires that exactly one `flash-*` and one `ram-*` feature is enabled. They are
  enabled automatically when using an `mcu-*` feature, but if you manually selected the other
  features before this will break the build because of the missing features.
- The `rtc::RTC` struct has been renamed to `rtc::Rtc` and includes a big refactoring and a few API
  changes. It now implements the traits from the [rtcc crate](https://docs.rs/rtcc/) and uses
  date/time types from Chrono.
- When downgrading GPIO pins, not only the pin number but also the port identifier is erased. To
  update your code, replace `PA`/`PB`/`PC`/`PD`/`PE`/`PH` with `Pin`. ([#190])

### Non-Breaking Changes

- The crate now has an optional `rtc` feature (enabled by default). The `rtc` module is only
  available if that feature is enabled. Enabling the feature also pulls in the `rtcc` and `chrono`
  dependencies, in oder to support a richer calendar / clock API.

### Fixes

- Fixed potential race condition when flushing the tx serial buffer.
- Fixed RTC year handling. Previously, the implementation incorrectly assumed that the BCD year 00
  corresponds to 1970, but this results in a wrong leap year calculation. The correct time base is
  the year 2000.

## [v0.7.0] - 2021-03-10

### Additions

- Timers: Add support for TIM6 ([#101])
- Timers: Implement `LinkedTimer` ([#115]). It is initialized with two hardware timers (either
  TIM2/TIM3 or TIM21/TIM22). The two timers are configured in master/slave mode so that an overflow
  of the master timer triggers an update on the slave timer. This way, two 16 bit timers can be
  combined to a single 32 bit timer. (The STM32L0 does not have 32 bit timers.)
- Remove config tests that prevented stm32l0x1 devices from using ADC with DMA ([#124])
- Add `enable_lse` method to RCC structure that configures LSE ([#130])
- Expose factory calibration data ([#121])
- Prelude: add `Serial1Ext` ([#133])
- Flash: support stm32l0x3 ([#134])
- Add Cargo features for EEPROM size, generated with cube-parse ([#137])
- Flash: Support writing single bytes to EEPROM ([#140])
- Add MCO support ([#143])
- Timers: Add QEI support for LPTIM ([#144])
- Timers: Add encoder input support for TIM2 and TIM21 ([#145])
- Timers: Adds support for LSE clocking to LPUART ([#131])
- SPI: Add DMA support ([#148])

### Breaking Changes

### Non-Breaking Changes

- Improvements to polling i²c driver ([#102])
- Enforce rustfmt ([#107])
- Simplify `flash.c` ([#136])

### Fixes

- I2C: Before starting new transaction ensure that both TX and RX buffers are empty ([#98])
- Detect rollover using positions rather than DMA flags ([#127]). This fixes a faulty DMA buffer
  overflow in `adc_trig` example ([#104]).
- Maximum frequency for stm32l0 increased to 32MHz ([#129])

build.rs
use std::env;
use std::fs::{self, File};
use std::io::Write;
use std::path::PathBuf;

fn main() {
    // Put the linker script somewhere the linker can find it
    let out = &PathBuf::from(env::var_os("OUT_DIR").unwrap());

    let mut feature_count = 0;

    if cfg!(feature = "stm32l0x1") {
        feature_count += 1;
    }

    if cfg!(feature = "stm32l0x2") {
        feature_count += 1;
    }

    if cfg!(feature = "stm32l0x3") {
        feature_count += 1;
    }

    if !cfg!(feature = "disable-linker-script") {
        if feature_count != 1 {
            panic!("\n\nMust select exactly one package for linker script generation!\nChoices: 'stm32l0x1' or 'stm32l0x2' or 'stm32l0x3'\nAlternatively, pick the mcu-feature that matches your MCU, for example 'mcu-STM32L071KBTx'\n\n");
        }

        let flash_features: Vec<u32> = [
            (8, cfg!(feature = "flash-8")),
            (16, cfg!(feature = "flash-16")),
            (32, cfg!(feature = "flash-32")),
            (64, cfg!(feature = "flash-64")),
            (128, cfg!(feature = "flash-128")),
            (192, cfg!(feature = "flash-192")),
        ]
        .iter()
        .filter(|(_, f)| *f)
        .map(|(f, _)| *f)
        .collect();

        if flash_features.len() != 1 {
            panic!("\n\nMust select exactly one flash size for linker script generation!\n\
            Choices: 'flash-8', 'flash-16', 'flash-32', 'flash-64', 'flash-128' or 'flash-192'\n \
            Alternatively, pick the mcu-feature that matches your MCU, for example 'mcu-STM32L071KBTx'\n\n");
        }

        let flash_size = flash_features[0];

        let ram_features: Vec<u32> = [
            (2, cfg!(feature = "ram-2")),
            (8, cfg!(feature = "ram-8")),
            (20, cfg!(feature = "ram-20")),
        ]
        .iter()
        .filter(|(_, f)| *f)
        .map(|(f, _)| *f)
        .collect();

        if ram_features.len() != 1 {
            panic!("\n\nMust select exactly one ram size for linker script generation!\n\
            Choices: 'ram-2', 'ram-8' or 'ram-20'\n \
            Alternatively, pick the mcu-feature that matches your MCU, for example 'mcu-STM32L071KBTx'\n\n");
        }

        let ram_size = ram_features[0];

        let linker = format!(
            r#"MEMORY
{{
    FLASH : ORIGIN = 0x08000000, LENGTH = {}K
    RAM : ORIGIN = 0x20000000, LENGTH = {}K
}}"#,
            flash_size, ram_size
        );

        File::create(out.join("memory.x"))
            .unwrap()
            .write_all(linker.as_bytes())
            .unwrap();
        println!("cargo:rustc-link-search={}", out.display());
    }

    println!("cargo:rerun-if-changed=build.rs");

    // Copy the binary blob required by the Flash API somewhere the linker can
    // find it, and tell Cargo to link it.

    let blob_name = "flash";
    let blob_file = format!("lib{}.a", blob_name);
    let blob_path = format!("flash-code/{}", blob_file);

    fs::copy(&blob_path, out.join(blob_file)).expect("Failed to copy binary blob for Flash API");

    println!("cargo:rustc-link-lib=static={}", blob_name);
    println!("cargo:rustc-link-search={}", out.display());

    println!("cargo:rerun-if-changed={}", blob_path);
}
